runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp

```diff
@@ -120,10 +121,58 @@ void J9::ARM64::JNILinkage::checkForJNIExceptions(TR::Node *callNode, TR::Regist
    }
 
 TR::Instruction *J9::ARM64::JNILinkage::generateMethodDispatch(TR::Node *callNode, bool isJNIGCPoint,
-                                                               TR::RegisterDependencyConditions *deps, uintptrj_t targetAddress)
+                                                               TR::RegisterDependencyConditions *deps, uintptrj_t targetAddress, TR::Register *scratchReg)
    {
-   TR_UNIMPLEMENTED();
-   return NULL;
+   TR::ResolvedMethodSymbol *resolvedMethodSymbol = callNode->getSymbol()->castToResolvedMethodSymbol();
+   TR_ResolvedMethod *resolvedMethod = resolvedMethodSymbol->getResolvedMethod();
+   TR_J9VMBase *fej9 = reinterpret_cast<TR_J9VMBase *>(fe());
+
+   TR::Instruction *cursor = cg()->getAppendInstruction();
+
+   if (fej9->needClassAndMethodPointerRelocations())
+      {
+      // load a 64-bit constant into a register with a fixed 4 instruction sequence
+      TR::Instruction *firstInstruction;
+
+      cursor = firstInstruction = generateTrg1ImmInstruction(cg(), TR::InstOpCode::movzx, callNode, scratchReg, targetAddress & 0x0000ffff, cursor);
+      cursor = generateTrg1ImmInstruction(cg(), TR::InstOpCode::movkx, callNode, scratchReg, ((targetAddress >> 16) & 0x0000ffff) | TR::MOV_LSL16, cursor);
+      cursor = generateTrg1ImmInstruction(cg(), TR::InstOpCode::movkx, callNode, scratchReg, ((targetAddress >> 32) & 0x0000ffff) | TR::MOV_LSL32, cursor);
+      cursor = generateTrg1ImmInstruction(cg(), TR::InstOpCode::movkx, callNode, scratchReg, (targetAddress >> 48) | TR::MOV_LSL48, cursor);
+
+      TR_ExternalRelocationTargetKind reloType;
+      if (resolvedMethodSymbol->isSpecial())
+         reloType = TR_JNISpecialTargetAddress;
+      else if (resolvedMethodSymbol->isStatic())
+         reloType = TR_JNIStaticTargetAddress;
+      else if (resolvedMethodSymbol->isVirtual())
+         reloType = TR_JNIVirtualTargetAddress;
+      else
+         {
+         reloType = TR_NoRelocation;
+         TR_ASSERT(0, "JNI relocation not supported.");
+         }
+      cg()->addExternalRelocation(new (trHeapMemory()) TR::BeforeBinaryEncodingExternalRelocation(
+                                                            firstInstruction,
+                                                            reinterpret_cast<uint8_t *>(callNode->getSymbolReference()),
+                                                            reinterpret_cast<uint8_t *>(callNode->getInlinedSiteIndex()),
+                                                            reloType, cg()),
+                                                          __FILE__,__LINE__, callNode);
+
+      }
+   else
+      {
+      loadConstant64(cg(), callNode, targetAddress, scratchReg);
+      }
+   // Add the first instruction of address materialization sequence to JNI call sites
+   cg()->getJNICallSites().push_front(new (trHeapMemory()) TR_Pair<TR_ResolvedMethod, TR::Instruction>(resolvedMethod, cursor->getNext()));
```

Comment:
What does `cursor` point to here?
There is a mismatch between the `needClassAndMethodPointerRelocations()` path and the `else` path.

Reply:
`cursor` should point to the first instruction of the method address materialization sequence.
I updated the code.

Reply:
I noticed that `JNICallSite` is patched for `registerNatives`/`unregisterNatives`.
I changed that method to use the fixed address materialization sequence so that it is always patchable. Would you review this again? Thanks.